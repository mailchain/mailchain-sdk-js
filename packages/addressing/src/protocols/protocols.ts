import { isAnyHex } from '@mailchain/encoding';
import { ETHEREUM, MAILCHAIN, ProtocolType } from './consts';

/**
 * Tries to guess what protocols this address could be. It will check for length, characters, and prefixes (where relevant).
 * @param address
 * @returns List of possible protocols. There are no guarantees the address exists on any of the returned protocols.
 */
export function guessProtocolsFromAddress(address: string): ProtocolType[] {
	address = address.trim();
	const isAddressHex = isAnyHex(address);

	const protocols = new Set<ProtocolType>();
	// needs to be a series of if statements as can be multiple protocols
	if (isAddressHex && address.length === 42 && address.startsWith('0x')) {
		protocols.add(ETHEREUM);
	}
	if (address.length <= 20 && address.length >= 3 && new RegExp('^[a-zA-Z0-9_-]+$').test(address)) {
		protocols.add(MAILCHAIN);
	}

	return Array.from(protocols);
}
