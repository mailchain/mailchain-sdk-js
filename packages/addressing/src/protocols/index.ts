export * from './consts';
export * from './errors';
export * from './protocols';
