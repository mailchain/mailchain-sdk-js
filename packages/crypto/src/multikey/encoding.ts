import { PrivateKey, PublicKey } from '..';
import { IdFromPrivateKey, IdFromPublicKey, PrivateKeyFromId, PublicKeyFromId } from './ids';

export function EncodePublicKey(key: PublicKey): Uint8Array {
	const out = new Uint8Array(key.bytes.length + 1);

	out[0] = IdFromPublicKey(key);
	out.set(key.bytes, 1);

	return out;
}

export function DecodePublicKey(encoded: Uint8Array): PublicKey {
	if (encoded.length < 32 + 1) {
		throw Error('encoded public key is too short');
	}
	return PublicKeyFromId(encoded[0], encoded.slice(1));
}

export function EncodePrivateKey(key: PrivateKey): Uint8Array {
	const out = new Uint8Array(key.bytes.length + 1);

	out[0] = IdFromPrivateKey(key);
	out.set(key.bytes, 1);

	return out;
}

export function DecodePrivateKey(encoded: Uint8Array): PrivateKey {
	if (encoded.length < 32 + 1) {
		throw Error('encoded private key is too short');
	}

	return PrivateKeyFromId(encoded[0], encoded.slice(1));
}
