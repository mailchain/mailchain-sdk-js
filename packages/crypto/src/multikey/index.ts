export * from './names';
export * from './bytes';
export * from './ids';
