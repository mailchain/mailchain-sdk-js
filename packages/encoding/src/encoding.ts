import { decodeBase32, encodeBase32 } from './base32';
import { decodeBase58, encodeBase58 } from './base58';
import { encodeBase64UrlSafe } from './base64';
import { EncodingTypes, EncodingType } from './consts';
import { decodeHex, decodeHexZeroX, encodeHex, encodeHexZeroX } from './hex';
import { decodeUtf8, encodeUtf8 } from './utf8';

class UnsupportedEncodingError extends Error {
	constructor(encoding: string) {
		super(`Encoding [${encoding}] not supported`);
	}
}

/**
 * Decode returns the bytes represented by the decoded string src.
 * Decode uses the decode method mapped to kind parameter.
 *
 * If the input is kind is unknown or the input is malformed for the decode method it returns an error.
 * @param encoding
 * @param src
 * @returns
 */
export function decode(encoding: EncodingType, src: string): Uint8Array {
	switch (encoding.toLowerCase()) {
		case EncodingTypes.Base58:
			return decodeBase58(src);
		case EncodingTypes.Hex:
			return decodeHex(src);
		case EncodingTypes.Hex0xPrefix:
			return decodeHexZeroX(src);
		case EncodingTypes.Utf8:
			return decodeUtf8(src);
		case EncodingTypes.Base32:
			return decodeBase32(src);
		default:
			throw new UnsupportedEncodingError(encoding);
	}
}

/**
 * Encode returns the bytes encoded as requested by the encoding parameter.
 * @param encoding
 * @param src
 * @returns encoded value
 */
export function encode(encoding: EncodingType, src: Uint8Array): string {
	switch (encoding.toLowerCase()) {
		case EncodingTypes.Base58:
			return encodeBase58(src);
		case EncodingTypes.Hex:
			return encodeHex(src);
		case EncodingTypes.Hex0xPrefix:
			return encodeHexZeroX(src);
		case EncodingTypes.Base64:
			return encodeBase64UrlSafe(src);
		case EncodingTypes.Utf8:
			return encodeUtf8(src);
		case EncodingTypes.Base32:
			return encodeBase32(src);
		default:
			throw new UnsupportedEncodingError(encoding);
	}
}
