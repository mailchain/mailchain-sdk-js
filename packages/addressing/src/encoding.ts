import { EncodingTypes, EncodingType, decode, encode } from '@mailchain/encoding';
import { ALGORAND, ALL_PROTOCOLS, ETHEREUM, MAILCHAIN, ProtocolType, SUBSTRATE } from './protocols';

/**
 * Convert address from Uint8Array, selects the relevant encoding method and encodes it as string.
 */
export function encodeAddressByProtocol(
	address: Uint8Array,
	protocol: ProtocolType,
): { encoded: string; encoding: string } {
	const encoding = encodingByProtocol(protocol);
	const encoded = encode(encoding, address);

	return { encoded, encoding };
}

/**
 * Convert address from string, selects the relevant encoding method and decodes it as Uint8Array.
 */
export function decodeAddressByProtocol(
	address: string,
	protocol: ProtocolType,
): { decoded: Uint8Array; encoding: string } {
	const encoding = encodingByProtocol(protocol);
	const decoded = decode(encoding, address);

	return { decoded, encoding };
}

/**
 * EncodingByProtocol returns the relevant encoding method the protocol commonly uses.
 */
//
export function encodingByProtocol(protocol: ProtocolType): EncodingType {
	switch (protocol) {
		case ALGORAND:
			return EncodingTypes.Base32;
		case ETHEREUM:
			return EncodingTypes.Hex0xPrefix;
		case SUBSTRATE:
			return EncodingTypes.Base58;
		case MAILCHAIN:
			return EncodingTypes.Utf8;
		default:
			throw new Error(`unknown address encoding of [${protocol}]. Supported ${ALL_PROTOCOLS}.`);
	}
}
