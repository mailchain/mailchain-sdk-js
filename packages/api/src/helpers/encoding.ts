import { EncodingType, EncodingTypes } from '@mailchain/encoding';
import { AddressEncodingEnum } from '../api/api';

export function encodingTypeToEncodingEnum(encoding: EncodingType): AddressEncodingEnum {
	switch (encoding) {
		case EncodingTypes.Hex:
			return AddressEncodingEnum.HexPlain;
		case EncodingTypes.Hex0xPrefix:
			return AddressEncodingEnum.Hex0xPrefix;
		case EncodingTypes.Utf8:
			return AddressEncodingEnum.TextUtf8;
		case EncodingTypes.Base58:
			return AddressEncodingEnum.Base58Plain;
	}
	throw new Error(`unsupported encoding by API of [${encoding}]`);
}
