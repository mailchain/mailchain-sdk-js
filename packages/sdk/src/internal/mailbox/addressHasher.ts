import { PublicKey, decodePublicKey } from '@mailchain/crypto';
import { decodeHexZeroX } from '@mailchain/encoding';
import { ProtocolType } from '@mailchain/addressing';
import { KeyRing } from '@mailchain/keyring';
import { sha3_256 } from '@noble/hashes/sha3';
import { AddressesApiFactory } from '../api';

export type AddressesHasher = (addresses: string[]) => Promise<{ [address: string]: Uint8Array[] | undefined }>;

export function mailchainAddressHasher(
	addressesApi: ReturnType<typeof AddressesApiFactory>,
	keyRing: KeyRing,
): AddressesHasher {
	return async (addresses: string[]) => {
		const addressesHashes: { [address: string]: Uint8Array[] } = {};
		for (const address of addresses) {
			const addressHashes = [] as Uint8Array[];
			try {
				const { identityKey, protocol } = await addressesApi
					.getAddressIdentityKey(address)
					.then((res) => res.data);
				const cryptoPublicKey = decodePublicKey(decodeHexZeroX(identityKey!));
				const hash = createAddressIdentityKeyHash(
					keyRing.accountIdentityKey().publicKey,
					cryptoPublicKey,
					protocol as ProtocolType,
				);
				addressHashes.push(hash);
			} catch (e) {
				console.warn(`failed creating AddressIdentityKey hash for address [${address}]`, e);
			}
			const hash = createAddressValueHash(keyRing.accountIdentityKey().publicKey, address.split('@')[0]);
			addressHashes.push(hash);
			addressesHashes[address] = addressHashes;
		}
		return addressesHashes;
	};
}

function createAddressIdentityKeyHash(
	accountIdentity: PublicKey,
	addressIdentity: PublicKey,
	protocol: ProtocolType,
): Uint8Array {
	const input = new Uint8Array([...addressIdentity.bytes, ...Buffer.from(protocol), ...accountIdentity.bytes]);
	return sha3_256(input).slice(0, 16); // truncate to 16bytes
}

function createAddressValueHash(accountIdentity: PublicKey, addressValue: string): Uint8Array {
	const input = new Uint8Array([...Buffer.from(addressValue), ...accountIdentity.bytes]);
	return sha3_256(input).slice(0, 16); // truncate to 16bytes
}
