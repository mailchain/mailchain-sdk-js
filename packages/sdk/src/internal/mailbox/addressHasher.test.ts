import {
	AliceED25519PrivateKey,
	AliceED25519PublicKey,
	BobED25519PublicKey,
} from '@mailchain/crypto/ed25519/test.const';
import { encodePublicKey } from '@mailchain/crypto/multikey/encoding';
import { encodeHexZeroX } from '@mailchain/encoding';
import { KeyRing } from '@mailchain/keyring';
import { ETHEREUM, MAILCHAIN } from '@mailchain/addressing/protocols';
import { mailchainAddressHasher } from './addressHasher';

test('addressHasher should create unique 128bit hash of addresses', async () => {
	const mockAddressesApi = { getAddressIdentityKey: jest.fn() };
	mockAddressesApi.getAddressIdentityKey
		.mockResolvedValueOnce({
			data: {
				identityKey: encodeHexZeroX(encodePublicKey(BobED25519PublicKey)),
				protocol: ETHEREUM,
			},
		})
		.mockResolvedValueOnce({
			data: {
				identityKey: encodeHexZeroX(encodePublicKey(BobED25519PublicKey)),
				protocol: MAILCHAIN,
			},
		})
		.mockResolvedValueOnce({
			data: {
				identityKey: encodeHexZeroX(encodePublicKey(AliceED25519PublicKey)),
				protocol: ETHEREUM,
			},
		});
	const keyRing = KeyRing.fromPrivateKey(AliceED25519PrivateKey);
	const hasher = mailchainAddressHasher(mockAddressesApi as any, keyRing);

	const firstRes = await hasher(['0x1234']);
	const differentProtocol = await hasher(['0x1234']);
	const differentIdentityKey = await hasher(['0x1234']);

	expect(firstRes).not.toEqual(differentProtocol);
	expect(firstRes).not.toEqual(differentIdentityKey);
	expect(differentProtocol).not.toEqual(differentIdentityKey);
	expect(firstRes).toMatchSnapshot('firstRes');
	expect(differentProtocol).toMatchSnapshot('differentProtocol');
	expect(differentIdentityKey).toMatchSnapshot('differentIdentityKey');
});

test('addressHasher should create only address hash when identity key hash not available', async () => {
	const mockAddressesApi = { getAddressIdentityKey: jest.fn() };
	mockAddressesApi.getAddressIdentityKey.mockRejectedValue(new Error('Failed identity key fetch'));
	const keyRing = KeyRing.fromPrivateKey(AliceED25519PrivateKey);
	const hasher = mailchainAddressHasher(mockAddressesApi as any, keyRing);

	const res = await hasher(['0x1234']);

	expect(res['0x1234']).toHaveLength(1);
	expect(res['0x1234']![0]).toMatchSnapshot('address hash');
});
