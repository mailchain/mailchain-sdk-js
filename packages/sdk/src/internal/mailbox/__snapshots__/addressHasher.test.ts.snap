// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addressHasher should create only address hash when identity key hash not available: address hash 1`] = `
Uint8Array [
  72,
  19,
  199,
  39,
  218,
  113,
  229,
  227,
  38,
  42,
  191,
  73,
  152,
  19,
  227,
  114,
]
`;

exports[`addressHasher should create unique 128bit hash of addresses: differentIdentityKey 1`] = `
Object {
  "0x1234": Array [
    Uint8Array [
      33,
      252,
      0,
      98,
      58,
      65,
      172,
      174,
      44,
      213,
      184,
      39,
      155,
      25,
      78,
      102,
    ],
    Uint8Array [
      72,
      19,
      199,
      39,
      218,
      113,
      229,
      227,
      38,
      42,
      191,
      73,
      152,
      19,
      227,
      114,
    ],
  ],
}
`;

exports[`addressHasher should create unique 128bit hash of addresses: differentProtocol 1`] = `
Object {
  "0x1234": Array [
    Uint8Array [
      44,
      148,
      96,
      62,
      71,
      94,
      120,
      208,
      150,
      176,
      93,
      173,
      120,
      10,
      63,
      82,
    ],
    Uint8Array [
      72,
      19,
      199,
      39,
      218,
      113,
      229,
      227,
      38,
      42,
      191,
      73,
      152,
      19,
      227,
      114,
    ],
  ],
}
`;

exports[`addressHasher should create unique 128bit hash of addresses: firstRes 1`] = `
Object {
  "0x1234": Array [
    Uint8Array [
      162,
      207,
      165,
      174,
      156,
      127,
      26,
      174,
      51,
      176,
      159,
      128,
      5,
      238,
      232,
      205,
    ],
    Uint8Array [
      72,
      19,
      199,
      39,
      218,
      113,
      229,
      227,
      38,
      42,
      191,
      73,
      152,
      19,
      227,
      114,
    ],
  ],
}
`;
